### PR TITLE
Fix node connection check

### LINKED ISSUE #9361: Node connection check fail to identify unqualified last elements in the connecton or object list
Issue overview
--------------

It was discovered in the node connect check for the flow stream connections (Check 10), the logic has an error that it will let an “unqualified” object to pass the check.

As a symptom, it could let an object is supposedly fail the Check 10 node-connection check to (unexpectedly) pass the check, if the object is listed as the last component in the node connection map. The defect file is supposed to give a severe warning (and fail the simulation) according to the intended purpose of Check 10. However, the check is unexpected passed due to the problem identified here. 

This is kind of a hidden issue and is relatively hard to see symptoms or impact in most cases. However, it is actually impacting nearly every simulation that has a node-connection in it. 

Currently there is no work-around for the problem. Code revision is needed to fix the issue.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): v22.1 and v9.6; but likely occurred in all past versions.
 - Problem identified while working on a help desk ticket 16143

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file [here](https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9361/9361_in_2only.idf))
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #9361 
 - Fixed the node connection check problem in that it would otherwise fail to detect an unqualified node/connection in Check 10 when the unqualified node/connection is at the very end of the queue.

Note that for the final merge of this PR to develop: it would be better to have the current PR(#9363) reviewed (or merged) before finally merging PR #9364. The reason is that the defect file for the current PR will no longer show the problem after the fix in PR #9364. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Review by mjwitte — APPROVED** [review]
I've tested this with the defect file, error shows up now. The overall logic of this check is not obvious, but the changes here are straightforward. The new unit test fails with develop, and passes here. This is good to go - please update the branch one last time.

**Comment by mjwitte** [comment]
@Myoldmopar Care to take a look before this merges?

File: src/EnergyPlus/BranchNodeConnections.cc [hunk 1]
```diff
@@ -669,7 +669,7 @@ void CheckNodeConnections(EnergyPlusData &state, bool &ErrorsFound)
         FluidStreamInletCount.allocate(MaxFluidStream);
         FluidStreamOutletCount.allocate(MaxFluidStream);
         FluidStreamCounts.allocate(MaxFluidStream);
-        NodeObjects.allocate(state.dataBranchNodeConnections->NumOfNodeConnections);
+        NodeObjects.allocate(state.dataBranchNodeConnections->NumOfNodeConnections + 1);
```
Comment (Myoldmopar) [inline src/EnergyPlus/BranchNodeConnections.cc:672]: So it just always allocates the one extra, whereas before it would miss the last one?

Comment (jcyuan2020) [reply src/EnergyPlus/BranchNodeConnections.cc:672]: Partially yes. The actual key to the solution is the new added at Line 694:
```
        NodeObjects(NumObjects) = state.dataBranchNodeConnections->NumOfNodeConnections + 1;
```

That will let the loop check go till the end of useful range in `NodeObjects`. In most cases, the final value of `NumObjects` will be always smaller than `state.dataBranchNodeConnections->NumOfNodeConnections` since there usually always objects that having two or more nodes, which will then be "folded" into the same `NodeObject(i)' element. So `state.dataBranchNodeConnections->NumOfNodeConnections` is usually quite enough as a valid the upper limit. 

The extra allocation (+1) is to ensure in all cases, the allocated length is still enough for all scenarios, since the new Line 694 will indeed do value assignment on an extra element which needs to be assumed to be always allocated. This is usually not a problem for most normal cases where there are objects that having more than 1 nodes so the original allocation length is usually quite enough. However, for some simple corner cases (e.g. in one of the existing unit tests), out-of-bound segment fault could happen if an extra allocation is not added. The `+1` allocation is a needed safety measure for such corner cases.

Comment (Myoldmopar) [reply src/EnergyPlus/BranchNodeConnections.cc:672]: Well, that definitely doesn't make me feel better about that code as a whole :).  I would like to think that we could identify exactly how big this array needs to be based on _something_.... but it seems your fix is better than what is in develop, so this is fine for right now.  It would be good to identify what is missing that is causing this uncertainty, and language such as: "usually quite enough".  Is it because we are trying to allocate this array too early, and we should defer this step until we have the objects set up?  Or is it because this array should grow dynamically each time a new object is identified rather than trying to guess the size up front?  Anyway, for the scope of this defect fix, it seems this change gets it moving, but it would be nice to go further on a follow up PR.

Comment (jcyuan2020) [reply src/EnergyPlus/BranchNodeConnections.cc:672]: I agree that growing dynamically might be the perfect (and neatest) way for memory allocation. However, it is might not be the best for efficiency, given that there would be repeated 1-increament on array size along the way in the loop. So it looks the current way is somehow a balanced mid-point between perfect memory size vs. repeated vector-size increment operations. 

On the other hand, although named like a complex "node object" array/vector, that particular `NodeObjects` array is actually just of a local primitive int type (Array1D int) and the elements in it will be fully de-allocated even before the Check 10 block finishes. This probably already kept the memory impact to a minimal (size and scope wise). I would be more worried if it is indeed a Node "data struct" type array instead.

Comment (Myoldmopar) [reply src/EnergyPlus/BranchNodeConnections.cc:672]: > there would be repeated 1-increament on array size along the way

`std::vector` is intelligent and reserves more memory behind the scenes, dynamically growing without us knowing it, so that repeated `push_back` calls don't all actually cause a new memory allocation, just an adjustment of the apparent size that we see.

All that to say that this is fine, I'm marking it approved.  Thanks for talking it through.

**Review by Myoldmopar — APPROVED** [review]
